Sources/Main.cpp

```diff
@@ -370,6 +465,52 @@ const UnicodeString FlagNames[12] = {L"ПАЛУ3", L"!СДЛ1", L"!СДП1", L"!
 const UnicodeString AltFlagNames[12] = {L"П", L"!СДЛ1", L"!СДП1", L"!СДЛ2", L"!СДП2",
     L"РРР0", L"РРР3", L"А15", L"В15", L"П0", L"П1", L"П2"};
 
+bool TX584Form::ParseInput(UnicodeString str, unsigned &Number)
+{
+    int pos = 1;
+    UnicodeString token = NextWord(str, pos);
+    if (token != L"INPUT" && token != L"ВВОД") {
+        return false;
+    }
+
+    UnicodeString NumberString = str.SubString(pos, str.Length()-pos+1);
+    UnicodeString WithoutSpaces = AnsiLowerCase(ReplaceStr(NumberString.Trim(), L" ", L""));
+
+    Number = 0;
+    if (WithoutSpaces.Length() >= 16) {
+        // скорее всего, двоичные коды
```

Comment:
This parser is a bit lax. It basically allows any mix of binary digits and spaces. In #11 we discussed two supported binary formats: 16 binary digits without separators or 19 binary digits with 3 separators. We can easily parse these two formats using regular expressions. Please use the following code as a template for stricter parser implementation:

```cpp
TRegEx Binary16RegEx(L" *([01]{16})", TRegExOptions());
TRegEx Binary4x4RegEx(L" *([01]{4}) ([01]{4}) ([01]{4}) ([01]{4})", TRegExOptions());

if (Binary16RegEx.IsMatch(str, pos)) {
    // 16 двоичных цифр без разделителей
    TMatch Match = Binary16RegEx.Match(str, pos);
    UnicodeString Value = Match.Groups[1].Value;
    ...
    pos += Match.Length;
} else if (Binary4x4RegEx.IsMatch(str, pos)) {
    // 4 двоичных тетрады с разделителями
    TMatch Match = Binary4x4RegEx.Match(str, pos);
    for (int i = 1; i < Match.Groups.Count; i++) {
        UnicodeString Nibble = Match.Groups[i].Value;
        ...
    }
    pos += Match.Length;
} else {
    // 16-битное десятичное или шестнадцатиричное число
    UnicodeString NumberString = AnsiLowerCase(NextWord(str, pos));
    ...
}
```

I recommend to make a helper method that converts an arbitrary string of 0s and 1s to integer number. It can be called `BinaryToInt` or something like this.